include/ofi_atomic_queue.h

```diff
@@ -0,0 +1,186 @@
+	assert(size == roundup_power_of_two(size));		\
+	aq->size = size;					\
+	aq->size_mask = aq->size - 1;				\
+	ofi_atomic_initialize64(&aq->write_pos, 0);		\
+	ofi_atomic_initialize64(&aq->read_pos, 0);		\
+	for (i = 0; i < size; i++)				\
+		ofi_atomic_initialize64(&aq->entry[i].seq, i);	\
+}								\
+								\
+static inline struct name * name ## _create(size_t size)	\
+{								\
+	struct name *aq;					\
+	aq = (struct name*) calloc(1, sizeof(*aq) +		\
+		sizeof(struct name ## _entry) *			\
+		(roundup_power_of_two(size)));			\
+	if (aq)							\
+		name ##_init(aq, roundup_power_of_two(size));	\
+	return aq;						\
+}								\
+								\
+static inline void name ## _free(struct name *aq)		\
+{								\
+	free(aq);						\
+}								\
+static inline int name ## _next(struct name *aq,		\
+		entrytype **buf, int64_t *pos)			\
+{								\
+	struct name ## _entry *ce;				\
+	int64_t diff, seq;					\
+	*pos = atomic_load_explicit(&aq->write_pos.val,		\
+				    memory_order_relaxed);	\
+	for (;;) {						\
+		ce = &aq->entry[*pos & aq->size_mask];		\
+		seq = atomic_load_explicit(&(ce->seq.val),	\
+			memory_order_acquire);			\
+		diff = seq - *pos;				\
+		if (diff == 0) {				\
+			if (atomic_compare_exchange_weak(	\
+				&aq->write_pos.val, pos,	\
+				*pos + 1))			\
+				break;				\
+		} else if (diff < 0) {				\
+			return -FI_ENOENT;			\
+		} else {					\
+			*pos = atomic_load_explicit(		\
+				&aq->write_pos.val,		\
+				memory_order_relaxed);		\
+		}						\
+	}							\
+	*buf = &ce->buf;					\
+	return FI_SUCCESS;					\
+}								\
+static inline int name ## _head(struct name *aq,		\
+		entrytype **buf, int64_t *pos)			\
+{								\
+	int64_t diff, seq;					\
+	struct name ## _entry *ce;				\
+	*pos = atomic_load_explicit(&aq->read_pos.val,		\
+			memory_order_relaxed);			\
+	for (;;) {						\
+		ce = &aq->entry[*pos & aq->size_mask];		\
+		seq = ce->seq.val;				\
+		diff = seq - (*pos + 1);			\
+		if (diff == 0) {				\
+			if (atomic_compare_exchange_weak(	\
+				&aq->read_pos.val, pos,		\
+				*pos + 1))			\
+				break;				\
+		} else if (diff < 0) {				\
+			return -FI_ENOENT;			\
+		} else {					\
+			*pos = atomic_load_explicit(		\
+				&aq->read_pos.val,		\
+				memory_order_relaxed);		\
+		}						\
+	}							\
+	*buf = &ce->buf;					\
+	return FI_SUCCESS;					\
+}								\
+static inline void name ## _commit(entrytype *buf,		\
+				int64_t pos)			\
+{								\
+	struct name ## _entry *ce;				\
+	ce = container_of(buf, struct name ## _entry, buf);	\
+	atomic_store_explicit(&ce->seq.val, pos + 1,		\
+			      memory_order_release);		\
+}								\
+static inline void name ## _discard(struct name *aq,		\
+			entrytype *buf,				\
+			int64_t pos)				\
+{								\
+	struct name ## _entry *ce;				\
+	ce = container_of(buf, struct name ## _entry, buf);	\
+	atomic_store_explicit(&ce->seq.val,			\
+			      pos + aq->size,			\
+			      memory_order_release);		\
+}								\
+void dummy ## name (void) /* work-around global ; scope */
```

Comment:
The atomic queue should be in its own patch for review.  (In fact, from the commit message, this sounds like several changes merged into a single patch.)  A lock free queue is non-trivial, e.g. the ABA problem.  What is the origin of this implementation?  How can we validate that it isn't racy and have confidence that it does indeed work?

Reply:
A similar implementation exists in OPX provider as well as OpenMPI. I tested it with 16 processes running on the same node with osu_alltoall. I can split it again in its own patch. Initially, I had built my patch list into a set of patches which bring in the lock free implementation as the last step. It was incremental, such that I implemented the locking optimization first with modifying the current locking system that already existed, and then transitioning into the lockfree approach. This way, we have a history of the implementation steps. I prefer this breakdown as the history is important for future investigations if needed. If you'd like I can bring back this patch breakdown. Let me know what you prefer.

Reply:
If you can bring back that patch breakdown, that sounds better.  If this is derived from a similar implementation, we need to identify where that is along with the license.  OPX almost certainly borrowed it from somewhere else.  Do you have a link to the openmpi source?

Reply:
The original NEMISIS paper:
https://digital.library.unt.edu/ark:/67531/metadc873149/m2/1/high_res_d/881588.pdf

OpenMPI's Implementation: https://github.com/open-mpi/ompi/tree/main/opal/mca/btl/sm

OpenMPI's implementations has a few optimizations in it (and they also use all their own data structures):
1. They use a LIFO queue for the senders free queue's in order to provide more cache hits, and have the remote return the free queue's to the senders receive fifo queue
2. FastBoxes (skip queue for small messages)
3. Large message protocol
ect...

Reply:
Is this implemented in MPICH as well, then?